Sources/GRPCCore/Timeout.swift

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright 2023, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Dispatch
+
+#if canImport(Darwin)
+import Darwin
+#elseif canImport(Glibc)
+import Glibc
+#endif
+
+/// A timeout for a gRPC call.
+///
+/// It's a combination of an amount (expressed as an integer of at maximum 8 digits), and a unit, which is
+/// one of ``Timeout/Unit`` (hours, minutes, seconds, milliseconds, microseconds or nanoseconds).
+///
+/// Timeouts must be positive and at most 8-digits long.
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+@usableFromInline
+struct Timeout: CustomStringConvertible, Hashable, Sendable {
+  /// Possible units for a ``Timeout``.
+  internal enum Unit: Character {
+    case hours = "H"
+    case minutes = "M"
+    case seconds = "S"
+    case milliseconds = "m"
+    case microseconds = "u"
+    case nanoseconds = "n"
+  }
+
+  /// The largest amount of any unit of time which may be represented by a gRPC timeout.
+  static let maxAmount: Int64 = 99_999_999
+
+  /// The wire encoding of this timeout as described in the gRPC protocol.
+  /// See "Timeout" in https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#requests
+  let wireEncoding: String
+
+  @usableFromInline
+  let duration: Duration
```

Comment:
I considered getting the `duration` to be something that we'll want to do relatively often, so avoiding the extra allocations for each time we get it may have been something we were looking for. I can change it into a computed property though.